track/framebuffer.py

```diff
@@ -17,116 +17,166 @@
+
+    thermal = attr.ib()
+    filtered = attr.ib()
+    mask = attr.ib()
+    frame_number = attr.ib()
+    flow = attr.ib(default=None)
+    flow_clipped = attr.ib(default=False)
+    scaled_thermal = attr.ib(default=None)
+
+    @classmethod
+    def from_array(cls, frame_arr, frame_number, flow_clipped=False):
+        return cls(
+            frame_arr[TrackChannels.thermal],
+            frame_arr[TrackChannels.filtered],
+            frame_arr[TrackChannels.mask],
+            frame_number,
+            flow=frame_arr[TrackChannels.thermal],
+            flow_clipped=flow_clipped,
+        )
 
+    def as_array(self, split_flow=True):
+        if split_flow:
+            return [
+                self.thermal,
+                self.filtered,
+                self.flow[:, :, 0] if self.flow else None,
+                self.flow[:, :, 1] if self.flow else None,
+                self.mask,
+            ]
 
-class FrameBuffer:
-    """ Stores entire clip in memory, required for some operations such as track exporting. """
-
-    def __init__(self):
-        # all arrays should be np.float32
-        self.thermal = None
-        self.filtered = None
-        self.delta = None
-        self.mask = None
-        self.flow = None
-        self.reset()
-
-    @property
-    def has_flow(self):
-        return self.flow is not None and len(self.flow) != 0
-
-    def get_previous_filtered(self, region=None, frame_number=None):
-        if frame_number:
-            previous = frame_number - 1
-            if previous < 0:
-                return None
-        else:
-            previous = -2
-
-        if region:
-            return region.subimage(self.filtered[previous])
-        else:
-            return self.filtered[previous]
+        return [self.thermal, self.filtered, self.flow, self.mask]
 
-    def get_frame_channels(self, region, frame_number):
+    def generate_optical_flow(self, opt_flow, prev_frame, flow_threshold=40):
         """
-        Gets frame channels for track at given frame number.  If frame number outside of track's lifespan an exception
-        is thrown.  Requires the frame_buffer to be filled.
-        :param track: the track to get frames for.
-        :param frame_number: the frame number where 0 is the first frame of the track.
-        :return: numpy array of size [channels, height, width] where channels are thermal, filtered, u, v, mask
+        Generate optical flow from thermal frames
+        :param opt_flow: An optical flow algorithm
         """
+        height, width = self.thermal.shape
+        flow = np.zeros([height, width, 2], dtype=np.float32)
+        threshold = np.median(self.thermal) + flow_threshold
+        scaled_thermal = np.uint8(np.clip(self.thermal - threshold, 0, 255))
+        if prev_frame is not None:
+            # for some reason openCV spins up lots of threads for this which really slows things down, so we
+            # cap the threads to 2
+            cv2.setNumThreads(2)
```

Comment:
I wonder if this should be `1` on a pi or are we not using optical flow at all?

Reply:
not using it at all